source-mongodb/oplog.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+type oplogRecord struct {
+	Ts primitive.Timestamp `bson:"ts"`
+}
+
+const tsProperty = "ts"
+
+// Oplog time difference: the time difference is measured as the
+// difference between the latest and the oldest record in a MongoDB oplog, and
+// it is used as a measure of how long do oplog records stay in the oplog.
+// Note that this value can fluctuate depending on the density of oplog
+// records during a period, so continuously monitoring this value is necessary
+// and a single value is not representative
+func oplogTimeDifference(ctx context.Context, client *mongo.Client) (uint32, error) {
+	var db = client.Database("local")
+	var oplog = db.Collection("oplog.rs")
+
+	// get latest record
+	var opts = options.FindOne().SetSort(bson.D{{tsProperty, SortDescending}})
+	var latest oplogRecord
+	if err := oplog.FindOne(ctx, bson.D{}, opts).Decode(&latest); err != nil {
+		return 0, fmt.Errorf("querying latest oplog record: %w", err)
+	}
+
+	// get oldest record
+	opts = options.FindOne().SetSort(bson.D{{tsProperty, SortAscending}})
+	var oldest oplogRecord
+	if err := oplog.FindOne(ctx, bson.D{}, opts).Decode(&oldest); err != nil {
+		return 0, fmt.Errorf("querying oldest oplog record: %w", err)
+	}
+
+	return latest.Ts.T - oldest.Ts.T, nil
+}
+
+// Check whether the timestamp ts is included in the oplog. This check is
+// necessary to make sure that when we set StartAtOperationTime, that timestamp
+// is actually available in the oplog. If it is not available, the driver will
+// not error, and we will lose some events without knowing
+func oplogHasTimestamp(ctx context.Context, client *mongo.Client, ts time.Time) error {
+	var db = client.Database("local")
+	var oplog = db.Collection("oplog.rs")
+
+	// get oldest record
+	var opts = options.FindOne().SetSort(bson.D{{tsProperty, SortAscending}})
+	var oldest oplogRecord
+	if err := oplog.FindOne(ctx, bson.D{}, opts).Decode(&oldest); err != nil {
+		return fmt.Errorf("querying oldest oplog record: %w", err)
+	}
+	var oldestTs = time.Unix(int64(oldest.Ts.T), 0)
+
+	if oldestTs.After(ts) {
```

Comment:
Will this work if:
1. A backfill is started on a collection that has nothing in the oplog
2. During the backfill, there's a change on the table, resulting in an oplog entry with a timestamp after the backfill started
3. When this check runs, it will see  the oplog timestamp coming after the backfill started timestamp

I don't have much knowledge on how MongoDB oplogs work so maybe this isn't an issue at all, if hypothetically there are periodic "heartbeat" types of changes always recorded on some frequency or something like that.

Reply:
One final question (I think 😃 ): Will this cause an error if the oplog is completely empty? That could mean that nothing has ever been written to the oplog, which would be fine, and we wouldn't want to fail on that. Or it could mean that the record(s) we really need to make sure we captured are now gone, which we would want to fail on.

Reply:
> Will this work if:
 >   A backfill is started on a collection that has nothing in the oplog
    During the backfill, there's a change on the table, resulting in an oplog entry with a timestamp after the backfill started
    When this check runs, it will see the oplog timestamp coming after the backfill started timestamp

We will error in this case as well. I thought about this a bit. I think one way to resolve this is to not set `backfill_started_at` unless we have actually captured anything. If we have captured any documents during backfill, there _must_ be records in the oplog, and so we must be able to find the timestamp, however if we can't find anything to capture, it is possible that there is nothing in the oplog.

Reply:
Okay, actually on further investigation, it seems it is not possible to have an empty oplog: during the initialization process, MongoDB itself writes configuration for the replicaset, which generates some oplog records, so any MongoDB instance that has replicasets enabled (a pre-requisite for our capture), will have some oplog records available.

Reply:
> Also - is access to the oplog collection granted under the same circumstances as access to a change stream?

No, I think with this new change we require access to the oplog collection, so I'm adding a check for that 👍🏽